source/common/common/logger.h

```diff
@@ -430,6 +431,22 @@ template <Id id> class Loggable {
     }                                                                                              \
   } while (0)
 
+// This is to get us to pass the format check. We reference a real-world time source here.
+// I think it's not easy to mock/simulate time here as it stands today anyway.
+using t_logclock = std::chrono::steady_clock; // NOLINT
+
+#define ENVOY_LOG_PERIODIC(LEVEL, CHRONO_DURATION, ...)                                            \
+  do {                                                                                             \
+    static auto* last_hit = new std::atomic<uint64_t>();                                           \
+    auto last = last_hit->load();                                                                  \
+    const auto now = t_logclock::now().time_since_epoch().count();                                 \
+    if ((now - last) >                                                                             \
+            std::chrono::duration_cast<std::chrono::nanoseconds>(CHRONO_DURATION).count() &&       \
+        last_hit->compare_exchange_strong(last, now)) {                                            \
```

Comment:
It seems the `LEVEL` should be checked before restarting the timer. Otherwise there's a chance that any message might not get published into the logs at all.
Say, you call `ENVOY_LOG_PERIODIC()` frequently with a level that gets rejected by `ENVOY_LOG_COMP_LEVEL` (called later inside `ENVOY_LOG()`) - the timer restarts every time; and when you get to the msg on a level you'd actually want to publish into the logs, it might not get there due to the timer is still too small. At least if i get the intent correctly there. (Another way is to dedicate a timer for each `LEVEL`.)

Reply:
Similar comment to every other "sparse" logger there. There's a change no message would get published if the `LEVEL` isn't checked first.

Reply:
The log timer is local to a single `ENVOY_LOG` invocation in the code, so there is no mixture of levels.

Reply:
Ah, right, macros.